enhancements/etcd/etcd-size-tuning.md

```diff
@@ -0,0 +1,192 @@
+  - "@hasbro17, etcd team lead"
+api-approvers: # In case of new or modified APIs or API extensions (CRDs, aggregated apiservers, webhooks, finalizers). If there is no API change, use "None"
+  - None
+creation-date: 2024-01-24
+last-updated: 2024-01-30
+tracking-link: # link to the tracking ticket (for example: Jira Feature or Epic ticket) that corresponds to this enhancement
+  - https://issues.redhat.com/browse/ETCD-514
+---
+
+# etcd Size Tuning
+
+## Summary
+
+This enhancement would replace the hardcoded value for the etcd backend database quota with a human readable, configurable value; 8GiB.
+This enhancement only covers the mvp for a tech preview release of this new feature; a future enhancement will be necessary.
+
+## Motivation
+
+Customers have asked for the ability to increase the quota for the backend database to faciliate larger clusters.
+Specifically, in our build clusters, the number of namespaces and CRDs (and the size of the CRDs) have been increasing which drastically increase the amount of data stored in etcd.
+In 4.14, the perf team is now unable to run their benchmark suite.
+We want to be able to allow for this while minimizing the risk of them setting values that cause issues.
+
+### User Stories
+
+* As an administrator, I want to change the etcd backend size to decrease the risk of quota reached errors in very large clusters.
+* As an Openshift support, I want to walk a customer through changing the etcd size in a minimal number of steps.
+* As an etcd engineer, I want to easily add and test new sizes for performance impacts on both large and small clusters.
+
+### Goals
+
+* Add a configuration option using resource.Quantity to allow for human readable values; i.e. 16GiB.
+* Remove the parameters from the podspec rendering and replace with the value.
+* Add an API to allow admins to change the value.
+* Add blocker to disallow shrinking the size once set to a larger value.
+
+### Non-Goals
+
+* Handle consuming value changes without an etcd rollout.
+
+## Proposal
+
+Allow the customer/admin access to set the backend quota directly through a human readable value. The only restriction is that they are disallowed from setting the value lower than the current.
+
+The default value or unset value would be the current, hardcoded backend quota: 8GiB. This would allow for upgrades where this configuration value wouldn't nesscessarily be set.
+
+This would allow for more flexibility, but would also likely increase the number of support cases as the memory requirements will be higher: changing quota from 8GiB to 16GiB is not double the memory, it'll be 6x the memory since each etcd member could take up to 16GiB.
+We will need to work with the perf team to understand the memory requirements of largers quotas, and the effect of the quota on things like defragmentation.
```

Comment:
Didn't see an upper limit specified elsewhere so as part of the proposal we should specify an upper limit e.g 64 GB. 

With a note on the need to test with the perf scale team what a reasonable and safe upper bound would be before we GA this.

Also mention the upper limit in the API extensions section below so it's clear.

Reply:
Are there any other constraints on what values make sense? For example, does the limit need to be a power of 2? Is 8 the minimum?

Reply:
> does the limit need to be a power of 2

No limitation from etcd that I'm aware of. This limit just dictates the size of the DB file which is mapped into memory.

> Is 8 the minimum?

The upstream limit by default is 2GB https://etcd.io/docs/v3.5/dev-guide/limit/#storage-size-limit
However we've always configured this as 8GB by default for openshift.
We're not aiming to change the default yet, and we don't want to allow decreasing the limit so customers don't write lock the cluster if they accidentally reduce it to less than what's being used by their cluster.

For the upper limit, 64GB is a ballpark figure. Others have safely gone as high as 100GB.
https://www.cncf.io/blog/2019/05/09/performance-optimization-of-etcd-in-web-scale-data-scenario/

But there may be other constraints on the size of the DB file such as the time to restore, and transporting the snapshot to a new member could be slow depending on the hardware.

We'll have to test out those constraints on the upper limit that we choose here.